### PR TITLE
[Refactor] Re-organize the tasks and requirements

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Refactor codebase into data_type/tasks + refactor CI to follow the same paradigm.

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? [not needed for typos/docs]
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/custom_task.rst [hunk 1]
```diff
@@ -122,7 +122,7 @@ individually. These methods behave identically to PyTorch Lightning’s
 
 Here is the pseudo code behind :class:`~flash.core.model.Task` step:
 
-.. code:: python
+.. testcode:: custom_task
```
Comment (ethanwharris) [inline docs/source/custom_task.rst:125]: Presumably this testcode fails as it is meant to be pseudo-code?

File: docs/source/custom_task.rst [hunk 2]
```diff
@@ -161,34 +161,36 @@ generated ``dataset``.
...
     - ``default_data_source`` is the name of the data source to use by default when predicting.
-- Override the ``get_state_dict`` and ``load_state_dict`` methods. These methods are used to save and load your :class:`~flash.data.process.Preprocess` from a checkpoint.
+- Override the ``get_state_dict`` and ``load_state_dict`` methods. These methods are used to save and load your :class:`~flash.core.data.process.Preprocess` from a checkpoint.
 - Override the ``{train,val,test,predict}_default_transforms`` methods to specify the default transforms to use in each stage (these will be used if the transforms passed in the ``__init__`` are ``None``).
-    - Transforms are given as a mapping from hook name to callable transforms. You should use :class:`~flash.data.transforms.ApplyToKeys` to apply each transform only to specific keys in the data dictionary.
+    - Transforms are given as a mapping from hook name to callable transforms. You should use :class:`~flash.core.data.transforms.ApplyToKeys` to apply each transform only to specific keys in the data dictionary.
 
 .. testcode:: custom_task
 
+    from flash import Preprocess
```
Comment (ethanwharris) [inline docs/source/custom_task.rst:192]: Imports are defined in part 1, any reason to define them again here? We could just always import locally and remove part 1 to be consistent.

File: docs/source/custom_task.rst [hunk 3]
```diff
@@ -247,15 +249,17 @@ The recommended way to define a custom :class:`~flash.data.process.Preprocess` i
...
+:class:`~flash.core.data.data_source.DataSource` objects it supports, we just need to create a
+:class:`~flash.core.data.data_module.DataModule` which has a reference to the ``preprocess_cls`` we want it to use. For any
+data source whose name is in :class:`~flash.core.data.data_source.DefaultDataSources`, there is a standard
 ``DataModule.from_*`` method that provides the expected inputs. So in this case, there is the
-:meth:`~flash.data.data_module.DataModule.from_numpy` that will use our numpy data source.
+:meth:`~flash.core.data.data_module.DataModule.from_numpy` that will use our numpy data source.
 
 .. testcode:: custom_task
 
+    import flash
```
Comment (ethanwharris) [inline docs/source/custom_task.rst:261]: Same comment on imports as above

File: docs/source/custom_task.rst [hunk 4]
```diff
@@ -296,6 +302,8 @@ With a trained model we can now perform inference. Here we will use a few exampl
 
 .. testcode:: custom_task
 
+    model = RegressionTask(num_inputs=datamodule.train_dataset.num_inputs)
```
Comment (ethanwharris) [inline docs/source/custom_task.rst:305]: We already defined `model` in the previous block, any reason to repeat here?

File: docs/source/custom_task.rst [hunk 5]
```diff
@@ -276,6 +280,8 @@ supplying the task itself, and the associated data:
 
 .. testcode:: custom_task
 
+    from sklearn import datasets
```
Comment (ethanwharris) [inline docs/source/custom_task.rst:283]: Import comment again haha

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley: Just minor comments about `custom_task.rst`

File: flash/image/backbones.py [hunk 6]
```diff
@@ -22,16 +22,16 @@
 from pytorch_lightning import LightningModule
 from pytorch_lightning.utilities import _BOLTS_AVAILABLE, rank_zero_warn
 from torch import nn as nn
-from torchvision.models.detection.backbone_utils import resnet_fpn_backbone
 
 from flash.core.registry import FlashRegistry
-from flash.utils.imports import _TIMM_AVAILABLE, _TORCHVISION_AVAILABLE
+from flash.core.utilities.imports import _IMAGE_AVAILABLE, _TIMM_AVAILABLE, _TORCHVISION_AVAILABLE
```
Comment (edgarriba) [inline flash/image/backbones.py:27]: `_IMAGE_AVAILABLE` not used ?

File: flash/image/classification/data.py [hunk 7]
```diff
@@ -56,6 +62,7 @@ def __init__(
                 DefaultDataSources.FOLDERS: ImagePathsDataSource(),
                 DefaultDataSources.NUMPY: ImageNumpyDataSource(),
                 DefaultDataSources.TENSORS: ImageTensorDataSource(),
+                DefaultDataSources.TENSORS: ImageTensorDataSource(),
```
Comment (edgarriba) [inline flash/image/classification/data.py:65]: duplicated ?

File: flash/image/embedding/model.py [hunk 8]
```diff
@@ -39,13 +44,6 @@ class ImageEmbedder(Task):
         learning_rate: Learning rate to use for training, defaults to ``1e-3``.
         pooling_fn: Function used to pool image to generate embeddings, defaults to :func:`torch.max`.
 
-    Example:
```
Comment (edgarriba) [inline flash/image/embedding/model.py:42]: why is it removed ?

**Pull request merged**